app/src/org/commcare/tasks/UnzipTask.java

```diff
@@ -1,109 +1,137 @@
 package org.commcare.tasks;
 
-import android.util.Log;
+import android.net.Uri;
 
+import org.commcare.CommCareApplication;
 import org.commcare.tasks.templates.CommCareTask;
 import org.commcare.utils.FileUtil;
 import org.javarosa.core.io.StreamsUtil;
 import org.javarosa.core.services.Logger;
 import org.javarosa.core.services.locale.Localization;
 
-import java.io.BufferedInputStream;
 import java.io.BufferedOutputStream;
 import java.io.File;
+import java.io.FileInputStream;
+import java.io.FileNotFoundException;
 import java.io.FileOutputStream;
 import java.io.IOException;
-import java.util.Enumeration;
+import java.io.InputStream;
 import java.util.zip.ZipEntry;
-import java.util.zip.ZipFile;
+import java.util.zip.ZipInputStream;
 
 /**
  * @author ctsims
  */
 public abstract class UnzipTask<R> extends CommCareTask<String, String, Integer, R> {
     public static final int UNZIP_TASK_ID = 7212435;
 
-    public UnzipTask() {
+    protected UnzipTask() {
         this.taskId = UNZIP_TASK_ID;
-
         TAG = UnzipTask.class.getSimpleName();
     }
 
     @Override
     protected Integer doTaskBackground(String... params) {
-        File archive = new File(params[0]);
-        File destination = new File(params[1]);
+        if (params[0].startsWith("content://")) {
+            return unZipFromContentUri(params[0], params[1]);
+        } else {
+            return unZipFromFilePath(params[0], params[1]);
+        }
+    }
 
-        Log.d(TAG, "Unzipping archive '" + archive + "' to  '" + destination + "'");
+    private Integer unZipFromFilePath(String filePath, String destinationPath) {
+        File archive = new File(filePath);
+        ZipInputStream zis;
+        try {
+            zis = new ZipInputStream(new FileInputStream(archive));

```

Comment:
NIT but these two functions feel un-DRY (wet?) Once you have the `inputStream` the try/catch blocks and handling re identical